pd/src/state.rs

```diff
@@ -112,6 +109,34 @@ ON CONFLICT (id) DO UPDATE SET data = $1
             .await?;
         }
 
+        if block.is_epoch_boundary.is_none() {
+            return Err(anyhow::anyhow!("EndBlock must be called prior to Commit, `is_epoch_boundary` was not set on the pending block"));
+        }
+        if block.is_epoch_boundary.unwrap() {
+            // validator rates need updating on epoch boundaries
+            let validators = self.validators().await?;
+            for validator in validators {
+                tracing::info!("updating validator rates for validator: {:?}", validator.0);
+                // TODO @ava insert calls here
+                let validator_rate = 0;
+                let voting_power = 0;
+
+                let pubkey_str = serde_json::to_string(&validator.0)?;
+
+                if block.epoch.is_none() {
+                    return Err(anyhow::anyhow!("EndBlock must be called prior to Commit, `epoch` was not set on the pending block"));
+                }
+                query!(
+                r#" INSERT INTO validator_rates ( epoch, validator_pubkey, validator_rate, voting_power ) VALUES ($1, $2, $3, $4)"#,
+                block.epoch.as_ref().unwrap().0 as i64,
```

Comment:
nit: there's multiple places where `block.epoch` has to be handled, maybe it would be cleaner to get the inner value upfront once, like
```rust
let epoch = block.epoch.ok_or_else(|| {
    anyhow::anyhow!("EndBlock must be called prior to Commit, `epoch` was not set on the pending block")
})?;
```